src/server/enterprise/server/api_server.go

```diff
@@ -332,3 +346,237 @@ func (a *apiServer) Deactivate(ctx context.Context, req *ec.DeactivateRequest) (
+	}
+	if err := a.rollPachd(ctx, true); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	return &ec.PauseResponse{}, nil
+}
+
+const updatedAtFieldName = "pachyderm.com/updatedAt"
+
+// rollPachd changes pachds from paused to unpaused, or unpaused to paused.  It
+// does this by creating a ConfigMap with a MODE value, then updating the pachd
+// deployment spec’s template’s annotations with a new value, which causes
+// Kubernetes to terminate existing pods and start new ones.
+//
+// Since the ConfigMap is not managed by Helm, it will persist through Helm
+// operations: a cluster which is upgraded by Helm in a paused state will thus
+// remain paused.
+//
+// If the ConfigMap is already in the desired state, no changes are made; this
+// ensures that PauseStatus can do its checking appropriately.
+//
+// There is a special case in main to handle the case where there is no
+// ConfigMap and the '$(MODE)' reference is verbatim.
+func (a *apiServer) rollPachd(ctx context.Context, paused bool) error {
+	var (
+		kc        *kubernetes.Clientset = a.env.GetKubeClient()
+		namespace                       = a.env.Namespace
+	)
+	cc := kc.CoreV1().ConfigMaps(namespace)
+	c, err := cc.Get(ctx, "pachd-config", metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// if the ConfigMap is not found, then the cluster is by definition unpaused
+		if !paused {
+			return nil
+		}
+		// Since pachd-config is not managed by Helm, it may not exist.
+		c = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pachd-config",
+				Namespace: namespace,
+			},
+			Data: make(map[string]string),
+		}
+		c.Data["MODE"] = "paused"
+		c.Data["previous-mode"] = "full"
+		if _, err := cc.Create(ctx, c, metav1.CreateOptions{}); err != nil {
+			return errors.Errorf("could not create configmap: %w", err)
+		}
+	} else if err != nil {
+		return errors.Errorf("could not get configmap pachd-config: %w", err)
+	} else {
+		// Curiously, Get can return no error and an empty configmap!
+		// If this happens, need to set the name and namespace.
+		if c.ObjectMeta.Name == "" {
+			c.ObjectMeta = metav1.ObjectMeta{
+				Name:      "pachd-config",
+				Namespace: namespace,
+			}
+		}
+		if c.Data == nil {
+			c.Data = make(map[string]string)
+			c.Data["MODE"] = "full"
+			c.Data["previous-mode"] = "paused"
+		}
+		if paused && c.Data["MODE"] == "paused" {
+			// Short-circuit and do not update if configmap
+			// is already in the correct state.  This keeps
+			// the updatedAt semantics correct when checking
+			// the pause status.
+			return nil
+		} else if !paused && c.Data["MODE"] != "paused" {
+			// Short-circuit and do not update if configmap
+			// is already in the correct state.  This keeps
+			// the updatedAt semantics correct when checking
+			// the pause status.
+			return nil
+		}
+		if c.Data["previous-mode"] == "" {
+			return errors.Errorf("empty previous-mode in pachd-config ConfigMap")
+		}
+		mode := c.Data["MODE"]
```

Comment:
I think I understand my confusion here better. I think what was hard for me to follow is the use of "previous-mode" key as having an initial value of "paused" which isn't semantically true. I think it would be clearer if the logic here was explicit, i.e.:
`if paused then c.Data["previous-mode] = c.Data["MODE"]; c.Data["MODE"] = "paused"; else c.Data["MODE"] = originalMode`

Maybe if we were able to capture the original mode the semantics here would be simpler. What do you think?

Reply:
Would it help if rather than constructing a new ConfigMap twice in similar but inverted ways, there were a function to create a new base-state ConfigMap, and a function to invert it? I think that might make the logic clearer.